src/main/webapp/app/exercises/programming/shared/utils/programming-exercise.utils.ts

```diff
@@ -25,6 +28,35 @@ export const createBuildPlanUrl = (template: string, projectKey: string, buildPl
     }
 };
 
+export const createCommitUrl = (
+    template: string | undefined,
+    projectKey: string | undefined,
+    participation: Participation | undefined,
+    submission: ProgrammingSubmission,
+): string => {
+    const projectKeyLowerCase = projectKey?.toLowerCase();
+    let repoSlug: string | undefined = undefined;
+    if (participation?.type === ParticipationType.PROGRAMMING) {
+        const studentParticipation = participation as ProgrammingExerciseStudentParticipation;
+        if (studentParticipation.repositoryUrl) {
+            repoSlug = projectKeyLowerCase + '-' + studentParticipation.participantIdentifier;
+        }
+    } else if (participation?.type === ParticipationType.TEMPLATE) {
+        // In case of a test submisson, we need to use the test repository
+        repoSlug = projectKeyLowerCase + (submission?.type === SubmissionType.TEST ? '-tests' : '-exercise');
+    } else if (participation?.type === ParticipationType.SOLUTION) {
+        // In case of a test submisson, we need to use the test repository
+        repoSlug = projectKeyLowerCase + (submission?.type === SubmissionType.TEST ? '-tests' : '-solution');
+    }
+    if (repoSlug && template && projectKeyLowerCase) {
+        return template
+            .replace('{projectKey}', projectKeyLowerCase)
+            .replace('{repoSlug}', repoSlug)
+            .replace('{commitHash}', submission.commitHash ?? '');
+    }
+    return '';
```

Comment:
```suggestion
    const projectKeyLowerCase = projectKey?.toLowerCase();
    let repoSlug: string | undefined = undefined;
    if (participation?.type === ParticipationType.PROGRAMMING) {
        const studentParticipation = participation as ProgrammingExerciseStudentParticipation;
        if (studentParticipation.repositoryUrl) {
            repoSlug = studentParticipation.participantIdentifier;
        }
    } else if (participation?.type === ParticipationType.TEMPLATE) {
        // In case of a test submisson, we need to use the test repository
        repoSlug = (submission?.type === SubmissionType.TEST ? 'tests' : 'exercise');
    } else if (participation?.type === ParticipationType.SOLUTION) {
        // In case of a test submisson, we need to use the test repository
        repoSlug = (submission?.type === SubmissionType.TEST ? 'tests' : 'solution');
    }
    if (repoSlug && template && projectKeyLowerCase) {
        return projectKeyLowerCase + '-' + template
            .replace('{projectKey}', projectKeyLowerCase)
            .replace('{repoSlug}', repoSlug)
            .replace('{commitHash}', submission.commitHash ?? '');
    }
    return '';
```
I think this is equivalent. However, this seems a bit odd since you add the projectKey before the template and replace it in the template. Could the template here be changed to fit it better?

Reply:
I think I get what you mean, but this is not equivalent. You probably want `projectKeyLowerCase + '-' + repoSlug` in the replacement. I would then rename the variable, as it is no longer the repo slug but the postfix.

Reply:
Ah, yes. You are correct. My bad. But good change!